### PR TITLE
[DRAFT] Update Service to Dependency (Using properyWrapper)

### PR DESCRIPTION
**Pull request by evilpenguin**
Looking for feedback on this. Thoughts?

### CODE REVIEW

File: SayTheirNames/Source/View/Home/Person/PersonCell.swift [hunk 1]
```diff
@@ -103,7 +103,7 @@ final class PersonCell: UICollectionViewCell {
     func configure(with person: Person) {
         profileImageView.populate(withURL: person.images.first?.personURL ?? "")
         nameLabel.text = person.fullName.uppercased()
-        dateOfIncidentLabel.text = person.doi
+        dateOfIncidentLabel.text = person.doi.description
```
Comment (evilpenguin) [inline SayTheirNames/Source/View/Home/Person/PersonCell.swift:106]: Please ignore this, haha.

File: SayTheirNames/Source/Dependency/DependencyContainer.swift [hunk 2]
```diff
@@ -24,23 +24,63 @@
...
+// MARK: - Dependency
+protocol Dependency {
+    
 }
-/// This is a core class that holds all instances responsible for logic
-final class Service: Servicing {
+
+// MARK: - DependencyContainer
+final class DependencyContainer: Dependency {
+    lazy private(set) var navigator = Navigator()
```
Comment (hybridcattt) [inline SayTheirNames/Source/Dependency/DependencyContainer.swift:34]: can it be `let` ? the deps should be lightweight to create so I think it's ok if not lazy

File: SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift [hunk 3]
```diff
@@ -27,32 +27,33 @@ import Alamofire
 
 // MARK: - NetworkRequestor
 
-final class NetworkRequestor {
-    let concurrentQueue = DispatchQueue(label: "NetworkRequestor", attributes: .concurrent)
+final class NetworkRequestor: Dependency {
     let session: Session
-
+    let concurrentQueue = DispatchQueue(label: "NetworkRequestor", attributes: .concurrent)
+    private let headers = HTTPHeaders.init(["x-dates-epoch": "true"])
```
Comment (hybridcattt) [inline SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift:33]: could you make a separate PR for the date stuff, for visibility? 🙏

File: SayTheirNames/Source/Dependency/DependencyContainer.swift [hunk 4]
```diff
@@ -24,23 +24,63 @@
...
-    static let shared = Service()
-    
     // MARK: - Init
     init() {
         Log.mode = .all
         Log.print("SayTheirNames Version: \(Bundle.versionBuildString)")
         Log.print("Starting Services")
+        
+        // Handle our Injection
+        let injectionFactory = DependencyInjectionFactory.notShared
```
Comment (hybridcattt) [inline SayTheirNames/Source/Dependency/DependencyContainer.swift:46]: `notShared`? 🤔 😅  a bit confused, could you please explain?

File: SayTheirNames/Source/Dependency/DependencyContainer.swift [hunk 5]
```diff
@@ -24,23 +24,63 @@
...
+    func add<S: Dependency>(handle: S) {
+        let key = String(describing: handle.self)
+        self.factoryDict[key] = handle
+    }
+
+    func resolve<S: Dependency>(_ type: S.Type) -> S {
+        let key = String(reflecting: type)
+        let component = self.factoryDict[key]
+        
+        return component as! S
```
Comment (hybridcattt) [inline SayTheirNames/Source/Dependency/DependencyContainer.swift:84]: guard & fatal error would be better here (minor thing though)

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 2 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift]: We can indeed. Only thing is we cannot have a `Dependency` use `@DependencyInject`.

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 4 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift]: I mean it is shared but not realllly shared. Haha. Ill change it back.

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 2 above)
Comment (Kilo-Loco) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift]: I think as the amount of dependencies grows, using lazy would be alright so we aren't bringing these services into memory. Overall, I can go either way

File: SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift (hunk 3 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift:33]: Yup; gonna remove all of this logic into another PR for us.

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 5 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift]: Added

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 2 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift]: If we go lazy we can have a `Dependency`use `@DependencyInject`. Im ok with either our.

File: SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift (hunk 3 above)
Comment (evilpenguin) [reply SayTheirNames/Source/Dependency/Networking/NetworkRequestor.swift]: All moved.

File: SayTheirNames/Source/Dependency/DependencyContainer.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+// MARK: - DependencyContainer
+final class DependencyContainer: Dependency {
+    let navigator = Navigator()
+    let image = ImageService()
+    let dateFormatter = DateFormatterService()
+    let network = NetworkRequestor()
+    
+    // MARK: - Init
+    init() {
```
Comment (biscuitehh) [inline SayTheirNames/Source/Dependency/DependencyContainer.swift:40]: Would it be possible to pass in Navigator, ImageService, etc. in the initializer instead of instantiating them as properties? It's more of a preference thing, but it makes it a little less bound to those services and adds a little more visibility at the DependencyContainer call site. Also, if things like ImageService become protocols it'll make it way easier to swap in testing components for tests and alternative implementations.

**Comment by hybridcattt** [comment]
oooops didn't notice it's still a draft. merged! 🚀

File: SayTheirNames/Source/Dependency/DependencyContainer.swift (hunk 6 above)
Comment (biscuitehh) [reply SayTheirNames/Source/Dependency/DependencyContainer.swift:40]: FWIW it doesn't have to happen on this pull-request, but just a suggestion for the future :)
